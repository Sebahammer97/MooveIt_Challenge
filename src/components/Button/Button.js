import React, { Component } from "react";
import cx from "classnames";
import PropTypes from "prop-types";

class Button extends Component {
  // Definition of props types to this component
  static propTypes = {
    onClick: PropTypes.func,
    children: PropTypes.node,
    variant: PropTypes.string,
    className: PropTypes.string,
    label: PropTypes.string,
    size: PropTypes.string,
    disabledClassName: PropTypes.string,
    disabled: PropTypes.bool,
  };

  // Assigning default values for the props
  static defaultProps = {
    className: "",
    label: "Button",
    size: "",
    variant: "default",
    disabled: false,
    disabledClassName: "",
  };

  handleButtonClick = (event) => {
    const { onClick, disabled } = this.props;

    if (disabled) {
      return;
    }

    onClick &&
      onClick({
        event,
      });
  };

  renderChildren = () => {
    const { label, children } = this.props;

    if (label !== "Button") {
      return label;
    }

    if (children) {
      return children;
    }

    return "Button";
  };

  render() {
    const {
      className,
      size,
      variant,
      disabled,
      disabledClassName,
    } = this.props;

    const styles = require("./style.module.css");

    const _className = cx(
      className,
      styles[size],
      styles.button,
      styles[variant],
      {
        [styles.disabled]: disabled,
        [disabledClassName]: disabled,
      }
    );

    return (
      <div
        onClick={this.handleButtonClick}
        className={_className}
        data-testid="button"
      >
        {this.renderChildren()}
      </div>
    );
  }
}

export default Button;
