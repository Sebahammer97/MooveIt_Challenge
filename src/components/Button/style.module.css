.button {
  position: relative;
  font-size: 14px;
  font-weight: 600;
  text-align: center;
  padding: 0.7em 1.2em;
  cursor: pointer;
  user-select: none;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  height: 32px;
  min-width: 96px;
  border-radius: 4px;
  background-color: #fff;
  border: 1px solid;
  color: #fff;
  margin-bottom: 0.5em;
  -webkit-transition: background 0.2s ease;
  -moz-transition: background 0.2s ease;
  -o-transition: background 0.2s ease;
  transition: background 0.2s ease;
}

.small {
  font-size: 8px;
}

.medium {
  font-size: 16px;
}

.large {
  font-size: 24px;
}

.outline {
  background-color: #fff;
  border-color: #61dafb;
  color: #61dafb;
}

.outline:hover {
  background-color: #61dafb;
  border-color: #61dafb;
  color: #fff;
}

.default {
  background-color: #61dafb;
  border-color: #61dafb;
  color: #fff;
}

.default:hover {
  background-color: #fff;
  border-color: #fff;
  color: #61dafb;
}

.link {
  background-color: transparent;
  color: #4c4c4c;
  box-shadow: none;
}

.link:hover {
  color: #40a9ff;
  background-color: transparent;
}

.secondary {
  background-color: #fff;
  border-color: rgb(186, 186, 186);
  color: #aa66cd;
}

.secondary:hover {
  background-color: #aa66cd;
  border-color: #eee;
  color: #fff;
}

.primary {
  background-color: #2fcb53;
  border-color: #2fcb53;
  color: #fff;
}

.primary:hover {
  background-color: #48dd84;
  border-color: #48dd84;
}

.danger {
  color: #fff;
  background-color: rgb(214, 69, 64);
  border-color: rgb(214, 69, 64);
}

.danger:hover {
  background-color: rgb(208, 50, 45);
  border-color: rgb(208, 50, 45);
}

.disabled {
  background-color: #fff;
  border-color: #bababa;
  color: #bababa;
  box-shadow: none;
  opacity: 0.7;
  cursor: not-allowed;
  border: 1px solid;
}

.disabled:hover {
  background-color: #fff;
  color: #bababa;
  border-color: #bababa;
}

.reset {
  background-color: #fff;
  border-color: #bababa;
  color: #000;
  width: 20px;
  min-width: 0 !important;
  height: 20px;
  padding: 0 0;
  border-width: 0 0;
  margin: 0 0;
}
