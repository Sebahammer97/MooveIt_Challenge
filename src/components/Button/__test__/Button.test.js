import React from "react";
import ReactDOM from "react-dom";
import Button from "../Button";

import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

afterEach(cleanup);

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Button />, div);
});

it("renders button correctly", () => {
  const { getByTestId } = render(<Button label="Click me" />);
  expect(getByTestId("button")).toHaveTextContent("Click me");
});

it("matches snapshots button", () => {
  const tree = renderer.create(<Button label="Click me" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots disabled", () => {
  const tree = renderer.create(<Button disabled={true} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots size", () => {
  const tree = renderer
    .create(
      <div>
        <Button size="small" /> <Button size="medium" /> <Button size="large" />
      </div>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches snapshots variant styles", () => {
  const tree = renderer
    .create(
      <div>
        <Button variant="primary" /> <Button variant="secondary" />{" "}
        <Button variant="link" /> <Button variant="danger" />
      </div>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
