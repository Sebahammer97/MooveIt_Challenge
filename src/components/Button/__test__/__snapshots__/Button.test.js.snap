// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots button 1`] = `
<div
  className="button default"
  data-testid="button"
  onClick={[Function]}
>
  Click me
</div>
`;

exports[`matches snapshots disabled 1`] = `
<div
  className="button default disabled "
  data-testid="button"
  onClick={[Function]}
>
  Button
</div>
`;

exports[`matches snapshots size 1`] = `
<div>
  <div
    className="small button default"
    data-testid="button"
    onClick={[Function]}
  >
    Button
  </div>
   
  <div
    className="medium button default"
    data-testid="button"
    onClick={[Function]}
  >
    Button
  </div>
   
  <div
    className="large button default"
    data-testid="button"
    onClick={[Function]}
  >
    Button
  </div>
</div>
`;

exports[`matches snapshots variant styles 1`] = `
<div>
  <div
    className="button primary"
    data-testid="button"
    onClick={[Function]}
  >
    Button
  </div>
   
  <div
    className="button secondary"
    data-testid="button"
    onClick={[Function]}
  >
    Button
  </div>
   
  <div
    className="button link"
    data-testid="button"
    onClick={[Function]}
  >
    Button
  </div>
   
  <div
    className="button danger"
    data-testid="button"
    onClick={[Function]}
  >
    Button
  </div>
</div>
`;
